web/src/lib/components/pages/record.tsx

```diff
@@ -268,11 +264,12 @@ export default class RecordPage extends Component<RecordProps, RecordState> {
     this.props.onRecordStop && this.props.onRecordStop();
   }
 
-  newSentenceSet() {
+  async newSentenceSet() {
     // If we don't have any sentences in our cache, fill it and try again.
     if (this.sentenceCache.length < SET_COUNT) {
       console.error('slow path for getting new sentences');
-      this.refillSentenceCache().then(this.newSentenceSet.bind(this));
+      await this.refillSentenceCache();
+      this.newSentenceSet();
```

Comment:
with async/await, we can neatly turn this recursive function into an iterative implementation:

```typescript
    // if we don't have enough sentences in our cache, fill it before continuing
    while (this.sentenceCache.length < SET_COUNT) {
      console.error('slow path for getting new sentences');
      await this.refillSentenceCache();
    }
```
(disclaimer: untested code)

Reply:
nice